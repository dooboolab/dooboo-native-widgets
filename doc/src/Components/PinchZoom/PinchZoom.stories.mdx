
<Meta title="Components/PinchZoom" />

# PinchZoom
This Component is a component which makes child components zoomable.

## Props

| Property             |      Required      | Type                           | Default               |
| -------------------- | :----------------: | ------------------------------ | --------------------- |
| style                |                    | `ViewStyle`                    | {}                    |
| children             |                    | `ReactNode`                    | null                  |
| blockNativeResponder |                    | `boolean`                      | `true`                |

### style
The PinchZoom is an Animated.View using panResponder to react for touch events. So, it need a style like common View component.

### children
It's just ReactNode type.


### blockNativeResponder
If this value is false, then 