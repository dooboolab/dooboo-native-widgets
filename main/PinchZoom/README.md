# Pinch Zoom

> Simple pinch zoom component for react-native.  

## Preview

![Aug-24-2020 16-14-55](https://user-images.githubusercontent.com/17980230/91014922-3c6a4f80-e625-11ea-86a4-75e88d284b46.gif)

## Installation
It may included to next version of dooboo-ui.
Next line is not working currently.

```sh
yarn add dooboo-ui
```

## Props

| Property             |      Required      | Type                           | Default               |
| -------------------- | :----------------: | ------------------------------ | --------------------- |
| style                |                    | `ViewStyle`                    |                       |
| children             | :white_check_mark: | `ReactNode`                    |                       |
| blockNativeResponder |                    | `boolean`                      | `true`                |

- blockNativeResponder

The PinchZoom consumes native gesture event. So, during it working, the parent component using gesture event (ScrollView, FlatList, etc ... ) may not work. 
If you set this value to false, then the parent component will work but it may prevent zoom action.

## Getting started

### Import

```tsx
import { PinchZoom } from 'dooboo-ui';
```

### Usage
Just wrap the children with PinchZoom!

```tsx
function PinchZoomImage(): React.ReactElement {
  const [width, setWidth] = React.useState<string | number>('100%');

  return <PinchZoom style={{ width, height: 200 }}>
    <Image
      style={{ width, height: 200, backgroundColor: '#fff' }}
      onLoad={({ nativeEvent: { source } }): void => {
        if (source.width && source.height) {
          setWidth(200 * source.width / source.height);
        }
      }}
      source={source}
      resizeMode={'contain'}
    />
  </PinchZoom>
}
```
