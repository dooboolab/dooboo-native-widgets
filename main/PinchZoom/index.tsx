import { Animated, PanResponder, PanResponderInstance, ViewStyle } from 'react-native';
import React, {
  PropsWithChildren, ReactElement, Ref, forwardRef,
  useCallback, useEffect, useImperativeHandle, useRef, useState,
} from 'react';
import {
  TouchPosition, Vector, VectorType,
  getClamppedVector,
  getOriginScaleTargetPosition,
  getTranslate,
} from './utils';

type Props = PropsWithChildren <{
  style?: ViewStyle,
  blockNativeResponder?: boolean,
  onScaleChanged?(value: number): void,
  onTranslateChanged?(valueXY: VectorType): void,
  onRelease?(): void,
}>

export interface PinchZoomRef {
  animatedValue: { scale: Animated.Value, translate: Animated.ValueXY },
  setOffsetValues(_:{ scale: number, translate: VectorType }): void,
}

function PinchZoom(props: Props, ref: Ref<PinchZoomRef>): ReactElement {
  const {
    children,
    style,
    blockNativeResponder = true, onScaleChanged, onTranslateChanged,
    onRelease,
  } = props;
  const touches = useRef([new TouchPosition(), new TouchPosition()]).current;
  const targetPosition = useRef(new Vector()).current;
  const layoutCenter = useRef(new Vector()).current;
  const scaleValue = useRef({ offset: 1, current: 1 }).current;
  const translateValue = useRef({ offset: new Vector(), current: new Vector() }).current;
  const scale = useRef(new Animated.Value(1)).current;
  const translate = useRef(new Animated.ValueXY(new Vector())).current;
  const release = (): void => {
    Animated.timing(scale, {
      useNativeDriver: true,
      toValue: 1,
    }).start();
    Animated.timing(translate, {
      useNativeDriver: true,
      toValue: new Vector(),
    }).start();
  };
  const createPanResponder = (): PanResponderInstance => {
    return PanResponder.create({
      onStartShouldSetPanResponder: () => true,
      onStartShouldSetPanResponderCapture: () => true,
      onMoveShouldSetPanResponder: () => true,
      onMoveShouldSetPanResponderCapture: () => true,
      onPanResponderGrant: () => {
        const [touch1, touch2] = touches;
        touch1.setOffset({ x: 0, y: 0 });
        touch2.setOffset({ x: 0, y: 0 });
        scaleValue.offset = scaleValue.current;
        translateValue.offset.set(translateValue.current);
      },
      onPanResponderMove: ({ nativeEvent }, gestureState) => {
        const [touch1, touch2] = touches;
        if (nativeEvent.touches.length === 2) {
          const secondEvent = nativeEvent.touches[1];
          if (touch2.offset.x === 0 && touch2.offset.y === 0) {
            touch1.setOffset({ x: nativeEvent.locationX, y: nativeEvent.locationY });
            touch2.setOffset({ x: secondEvent.locationX, y: secondEvent.locationY });
            targetPosition.set(
              getOriginScaleTargetPosition({
                currentPosition: touch1.offset.center(touch2.offset),
                layoutCenter,
                scale: scaleValue.offset,
                translate: translateValue.offset,
              }),
            );
          }
          touch1.setCurrent({ x: nativeEvent.locationX, y: nativeEvent.locationY });
          touch2.setCurrent({ x: secondEvent.locationX, y: secondEvent.locationY });
          scaleValue.current = Math.max(
            1,
            scaleValue.offset * touch1.current.distance(touch2.current) / touch1.offset.distance(touch2.offset),
          );
          scaleValue.current = Math.sqrt(scaleValue.current * scaleValue.offset);
          scale.setValue(scaleValue.current);
          translateValue.current.set(getTranslate({
            targetPosition,
            layoutCenter,
            scale: scaleValue.current,
          }));
          translate.setValue(translateValue.current);
        } else if (
          touch2.offset.x === 0 && touch2.offset.y === 0 &&
            nativeEvent.touches.length === 1
        ) {
          const maxValue = layoutCenter.multiply(scaleValue.current + 1);
          translateValue.current = getClamppedVector({
            vector: translateValue.offset.add({ x: gestureState.dx, y: gestureState.dy }),
            max: maxValue,
            min: maxValue.multiply(-1),
          });
          translate.setValue(translateValue.current);
        }
      },
      onPanResponderRelease: () => {
        onRelease && onRelease();
      },
      onPanResponderTerminationRequest: () => {
        return true;
      },
      onShouldBlockNativeResponder: () => {
        return blockNativeResponder;
      },
    });
  };

  const [panResponder, setPanResponder] = useState<PanResponderInstance>(createPanResponder());

  useEffect(() => {
    if (!onTranslateChanged) return;
    const id = translate.addListener(onTranslateChanged);
    return ():void => {
      translate.removeListener(id);
    };
  }, [onTranslateChanged]);

  useEffect(() => {
    if (!onScaleChanged) return;
    const id = scale.addListener(({ value }) => onScaleChanged(value));
    return ():void => {
      scale.removeListener(id);
    };
  }, [onScaleChanged]);

  useEffect(() => {
    release();
    setPanResponder(createPanResponder());
  }, [blockNativeResponder]);

  const setOffsetValues = useCallback(({ scale, translate }) => {
    scaleValue.offset = scale;
    translateValue.offset = new Vector(translate);
  }, [scaleValue, translateValue]);

  useEffect(() => {
    setOffsetValues({ scale: 0, translate: { x: 0, y: 0 } });
  }, [children]);

  useImperativeHandle(ref, () => ({
    animatedValue: { scale, translate },
    setOffsetValues,
  }));

  return <Animated.View
    testID="PINCH_ZOOM_CONTAINER"
    onLayout={({ nativeEvent: { layout } }): void => {
      layoutCenter.x = layout.width / 2;
      layoutCenter.y = layout.height / 2;
    }}
    style={[
      style,
      {
        transform: [
          {
            translateX: translate.x,
          },
          { translateY: translate.y },
          { scale },
        ],
      },
    ]}
    {...(panResponder?.panHandlers || {})}
  >
    {children}
  </Animated.View>;
}

export { PinchZoom };

export default forwardRef<PinchZoomRef, Props>(PinchZoom);
