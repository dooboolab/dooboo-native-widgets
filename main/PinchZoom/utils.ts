
export interface VectorType {
  x: number;
  y: number;
}

export class Vector implements VectorType {
  x: number;
  y: number;

  constructor(vector: VectorType = { x: 0, y: 0 }) {
    this.set(vector);
  }

  // istanbul ignore next
  toString(): string {
    return JSON.stringify({ x: this.x, y: this.y });
  }

  distance = (v1: Vector, v2: Vector = this): number => {
    const diffX = v1.x - v2.x;
    const diffY = v1.y - v2.y;
    return Math.sqrt(diffX ** 2 + diffY ** 2);
  }

  set = (vector: VectorType): void => {
    this.x = vector.x;
    this.y = vector.y;
  }

  add = (a: VectorType): Vector => {
    return new Vector({ x: this.x + a.x, y: this.y + a.y });
  }

  multiply = (n: number, b: VectorType = this): Vector => {
    return new Vector({ x: n * b.x, y: n * b.y });
  }

  center = (a: VectorType, b: VectorType = this): Vector => {
    return new Vector({ x: (a.x + b.x) / 2, y: (a.y + b.y) / 2 });
  }

  subtract = (a: VectorType): Vector => {
    return new Vector({ x: this.x - a.x, y: this.y - a.y });
  }
}

export class TouchPosition {
  offset: Vector;
  current: Vector;
  constructor({ offset = new Vector(), current = new Vector() }: {offset?: Vector, current?: Vector} = {}) {
    this.offset = offset;
    this.current = current;
  }

  setOffset = ({ x, y }: VectorType): void => {
    this.offset.x = x;
    this.offset.y = y;
  }

  setCurrent = ({ x, y }: VectorType): void => {
    this.current.x = x;
    this.current.y = y;
  }
}
