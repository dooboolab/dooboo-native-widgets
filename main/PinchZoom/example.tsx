import { Dimensions, FlatList, Image, ImageSourcePropType } from 'react-native';
import PinchZoom from './';
import React from 'react';
import styled from 'styled-components/native';

const Container = styled.View`
  flex: 1;
  padding: 10px;
`;

const ItemContainer = styled.View`
  padding: 10px 0;
  width: ${(): number => Dimensions.get('screen').width - 20}px;
  min-height: 200px;
`;

const TitleText = styled.Text`
  font-size: 16px;
`;

const ContentText = styled.Text`
  font-size: 14px;
`;

const ImageContainer = styled.View`
  background-color: #CCC;
  width: 100%;
  height: 200px;
  align-items: center;
`;

const images = [
  // IC_LOGO,
  { uri: 'https://images.unsplash.com/photo-1519335337423-a3357c2cd12e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2168&q=80' },
  { uri: 'https://www.housingwire.com/wp-content/uploads/2019/09/Purple-technology-data-internet-3.jpg' },
  { uri: 'https://images.unsplash.com/photo-1543007354-0bc99b7d096b?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&w=1000&q=80' },
  { uri: 'https://p.bigstockphoto.com/eIdTXLbqQilMs9xbjvcs_bigstock-Aerial-View-Of-Sandy-Beach-Wit-256330393.jpg' },
  { uri: 'https://avatars2.githubusercontent.com/u/7970947?v=3&s=460' },
];

function ImageItem({ source, title, content }
:{ source: ImageSourcePropType, title: string, content: string }): React.ReactElement {
  const [width, setWidth] = React.useState<string | number>('100%');

  return <ItemContainer>
    <TitleText>{title}</TitleText>
    <ContentText>{content}</ContentText>
    <ImageContainer>
      <PinchZoom style={{ width, height: 200 }}>
        <Image
          style={{ width, height: 200, backgroundColor: '#fff' }}
          onLoad={({ nativeEvent: { source } }): void => {
            if (source.width && source.height) {
              setWidth(200 * source.width / source.height);
            }
          }}
          source={source}
          resizeMode={'contain'}
        />
      </PinchZoom>
    </ImageContainer>
  </ItemContainer>;
}

function ImageList(): React.ReactElement {
  return <Container>
    <FlatList
      data={images}
      style={{ flex: 1 }}
      keyExtractor={ (item):string => item.uri}
      renderItem={({ item, index }): React.ReactElement =>
        <ImageItem
          source={item}
          title={`Image ${(index + 1)}`}
          content="This line is for the description of the image" />
      } />
  </Container>;
}

function Article(): React.ReactElement {
  return <PinchZoom style={{ width: '100%', backgroundColor: '#eee', padding: 10 }}>
    <TitleText>Article Title</TitleText>
    <ContentText style={{ minHeight: 100, textAlignVertical: 'center' }}>
      {
`This block is for the content of the article. \
PinchZoom works at only Image but also other views. So, let's zoom in with your fingers.
`
      }
    </ContentText>
  </PinchZoom>;
}

export {
  Article,
  ImageList,
};
