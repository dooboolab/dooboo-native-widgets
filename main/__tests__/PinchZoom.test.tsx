import * as React from 'react';
import { RenderResult, act, fireEvent, render } from '@testing-library/react-native';
import {
  closeGesture,
  getChangedDistanceRatio,
  getTwoFingerStartEndPositions,
  moveGesture,
  openGesture,
} from './capturedGesture';

import { ImageList } from '../PinchZoom/PinchZoom.example';
import { PanResponderCallbacks } from 'react-native';
import PinchZoom from '../PinchZoom';
import renderer from 'react-test-renderer';

const TEST_CONTAINER_WIDTH = 300;
const TEST_CONTAINER_HEIGHT = 200;
const TEST_CONTAINER_CENTER = { x: 150, y: 100 };

jest.useFakeTimers();

jest.mock('react-native/Libraries/Interaction/PanResponder', () => {
  return {
    create: (responderCallback: PanResponderCallbacks)
    : { panHandlers : { responderCallback:PanResponderCallbacks } } => {
      return {
        panHandlers: {
          responderCallback,
        },
      };
    },
  };
});

describe('[PinchZoom] of ImageList render', () => {
  it(' should renders without crashing', () => {
    const rendered: renderer.ReactTestRendererJSON | null = renderer
      .create(<ImageList PinchZoom={PinchZoom}/>)
      .toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  describe('ImageList interaction test', () => {
    const rendered: RenderResult = render(<ImageList PinchZoom={PinchZoom}/>);
    const pinchZoomContainer = rendered.getAllByTestId('PINCH_ZOOM_CONTAINER')[Math.floor(Math.random() * 5)];

    it(' should set the center of layout position when onLayout called', () => {
      act(() => {
        fireEvent.layout(pinchZoomContainer, {
          nativeEvent: {
            layout: {
              width: TEST_CONTAINER_WIDTH,
              height: TEST_CONTAINER_HEIGHT,
            },
          },
        });
      });
    });

    it(' should zoom in by openGesture', () => {
      act(() => {
        const callBacks = pinchZoomContainer.props.responderCallback;
        openGesture.forEach(({ name, nativeEvent, gestureState }) => {
          callBacks[name] && callBacks[name]({ nativeEvent }, gestureState);
        });
      });

      const { transform } = pinchZoomContainer.props.style;
      const scale = transform.find(({ scale }) => scale != null).scale;
      const translateX = transform.find(({ translateX }) => translateX != null).translateX;
      const translateY = transform.find(({ translateY }) => translateY != null).translateY;

      const { start, end } = getTwoFingerStartEndPositions(openGesture);
      const expectedScale = getChangedDistanceRatio({ start, end });

      expect(scale).toBeCloseTo(expectedScale);

      const zoomInPosition = { x: (start.x1 + start.x2) / 2, y: (start.y1 + start.y2) / 2 };

      expect((zoomInPosition.x - TEST_CONTAINER_CENTER.x) * scale)
        .toBeCloseTo(zoomInPosition.x - TEST_CONTAINER_CENTER.x - translateX);

      expect((zoomInPosition.y - TEST_CONTAINER_CENTER.y) * scale)
        .toBeCloseTo(zoomInPosition.y - TEST_CONTAINER_CENTER.y - translateY);
    });

    it(' should be moved by moveGesture when it zoomed in', () => {
      const { transform } = pinchZoomContainer.props.style;

      const prevScale = transform.find(({ scale }) => scale != null).scale;
      const prevTranslateX = transform.find(({ translateX }) => translateX != null).translateX;
      const prevTranslateY = transform.find(({ translateY }) => translateY != null).translateY;

      act(() => {
        const callBacks = pinchZoomContainer.props.responderCallback;
        moveGesture.forEach(({ name, nativeEvent, gestureState }) => {
          callBacks[name] && callBacks[name]({ nativeEvent }, gestureState);
        });
      });

      const { transform: changedTransform } = pinchZoomContainer.props.style;

      const scale = changedTransform.find(({ scale }) => scale != null).scale;
      const translateX = changedTransform.find(({ translateX }) => translateX != null).translateX;
      const translateY = changedTransform.find(({ translateY }) => translateY != null).translateY;

      const moveGestureStateList = moveGesture.filter(({ name }) => name === 'onPanResponderMove')
        .map(({ gestureState }) => gestureState);
      const lastMoveGestureState = moveGestureStateList[moveGestureStateList.length - 1];
      const maxMovableValue = {
        x: (TEST_CONTAINER_WIDTH * scale - TEST_CONTAINER_WIDTH) / 2,
        y: (TEST_CONTAINER_HEIGHT * scale - TEST_CONTAINER_HEIGHT) / 2,
      };

      expect(scale).toBeCloseTo(prevScale);
      expect(translateX).toBeCloseTo(
        Math.max(
          Math.min(prevTranslateX + lastMoveGestureState.dx, maxMovableValue.x),
          -maxMovableValue.x,
        ),
      );
      expect(translateY).toBeCloseTo(
        Math.max(
          Math.min(prevTranslateY + lastMoveGestureState.dy, maxMovableValue.y),
          -maxMovableValue.y,
        ),
      );
    });

    it(' should zoom out by closeGesture', () => {
      act(() => {
        const callBacks = pinchZoomContainer.props.responderCallback;
        closeGesture.forEach(({ name, nativeEvent, gestureState }) => {
          callBacks[name] && callBacks[name]({ nativeEvent }, gestureState);
        });
      });

      const { transform } = pinchZoomContainer.props.style;
      const scale = transform.find(({ scale }) => scale != null).scale;
      const translateX = transform.find(({ translateX }) => translateX != null).translateX;
      const translateY = transform.find(({ translateY }) => translateY != null).translateY;

      const openScale = getChangedDistanceRatio(getTwoFingerStartEndPositions(openGesture));
      const closeScale = getChangedDistanceRatio(getTwoFingerStartEndPositions(closeGesture));
      expect(openScale * closeScale).toBeLessThan(1);

      expect(scale).toEqual(1);
      expect(translateX).toEqual(0);
      expect(translateY).toEqual(0);
    });

    it(' should not be moved by moveGesture when its scale is 1', () => {
      act(() => {
        const callBacks = pinchZoomContainer.props.responderCallback;
        moveGesture.forEach(({ name, nativeEvent, gestureState }) => {
          callBacks[name] && callBacks[name]({ nativeEvent }, gestureState);
        });
      });
      const { transform } = pinchZoomContainer.props.style;
      const scale = transform.find(({ scale }) => scale != null).scale;
      const translateX = transform.find(({ translateX }) => translateX != null).translateX;
      const translateY = transform.find(({ translateY }) => translateY != null).translateY;

      expect(scale).toEqual(1);
      expect(translateX).toEqual(0);
      expect(translateY).toEqual(0);
    });
  });
});
