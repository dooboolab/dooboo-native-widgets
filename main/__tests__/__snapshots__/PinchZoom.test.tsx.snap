// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[PinchZoom] of ImageList render  should renders without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
      },
    ]
  }
>
  <ScrollView
    data={
      Array [
        Object {
          "uri": "https://images.unsplash.com/photo-1519335337423-a3357c2cd12e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2168&q=80",
        },
        Object {
          "uri": "https://www.housingwire.com/wp-content/uploads/2019/09/Purple-technology-data-internet-3.jpg",
        },
        Object {
          "uri": "https://images.unsplash.com/photo-1587628736664-fdc50efb57b7?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
        },
        Object {
          "uri": "https://p.bigstockphoto.com/eIdTXLbqQilMs9xbjvcs_bigstock-Aerial-View-Of-Sandy-Beach-Wit-256330393.jpg",
        },
        Object {
          "uri": "https://avatars2.githubusercontent.com/u/7970947?v=3&s=460",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flex": 1,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Array [
            Object {
              "maxWidth": 730,
              "minHeight": 200,
              "paddingBottom": 10,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 10,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
              },
            ]
          }
        >
          Image 1
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontSize": 14,
              },
            ]
          }
        >
          This line is for the description of the image
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#CCC",
                "height": 200,
                "width": "100%",
              },
              Object {
                "overflow": "hidden",
              },
            ]
          }
        >
          <View
            onLayout={[Function]}
            responderCallback={
              Object {
                "onMoveShouldSetPanResponder": [Function],
                "onMoveShouldSetPanResponderCapture": [Function],
                "onPanResponderGrant": [Function],
                "onPanResponderMove": [Function],
                "onPanResponderRelease": [Function],
                "onPanResponderTerminationRequest": [Function],
                "onShouldBlockNativeResponder": [Function],
                "onStartShouldSetPanResponder": [Function],
                "onStartShouldSetPanResponderCapture": [Function],
              }
            }
            style={
              Object {
                "height": 200,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "width": "100%",
              }
            }
            testID="PINCH_ZOOM_CONTAINER"
          >
            <Image
              onLoad={[Function]}
              resizeMode="contain"
              source={
                Object {
                  "uri": "https://images.unsplash.com/photo-1519335337423-a3357c2cd12e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2168&q=80",
                }
              }
              style={
                Object {
                  "backgroundColor": "#fff",
                  "height": 200,
                  "width": "100%",
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Array [
            Object {
              "maxWidth": 730,
              "minHeight": 200,
              "paddingBottom": 10,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 10,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
              },
            ]
          }
        >
          Image 2
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontSize": 14,
              },
            ]
          }
        >
          This line is for the description of the image
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#CCC",
                "height": 200,
                "width": "100%",
              },
              Object {
                "overflow": "hidden",
              },
            ]
          }
        >
          <View
            onLayout={[Function]}
            responderCallback={
              Object {
                "onMoveShouldSetPanResponder": [Function],
                "onMoveShouldSetPanResponderCapture": [Function],
                "onPanResponderGrant": [Function],
                "onPanResponderMove": [Function],
                "onPanResponderRelease": [Function],
                "onPanResponderTerminationRequest": [Function],
                "onShouldBlockNativeResponder": [Function],
                "onStartShouldSetPanResponder": [Function],
                "onStartShouldSetPanResponderCapture": [Function],
              }
            }
            style={
              Object {
                "height": 200,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "width": "100%",
              }
            }
            testID="PINCH_ZOOM_CONTAINER"
          >
            <Image
              onLoad={[Function]}
              resizeMode="contain"
              source={
                Object {
                  "uri": "https://www.housingwire.com/wp-content/uploads/2019/09/Purple-technology-data-internet-3.jpg",
                }
              }
              style={
                Object {
                  "backgroundColor": "#fff",
                  "height": 200,
                  "width": "100%",
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Array [
            Object {
              "maxWidth": 730,
              "minHeight": 200,
              "paddingBottom": 10,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 10,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
              },
            ]
          }
        >
          Image 3
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontSize": 14,
              },
            ]
          }
        >
          This line is for the description of the image
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#CCC",
                "height": 200,
                "width": "100%",
              },
              Object {
                "overflow": "hidden",
              },
            ]
          }
        >
          <View
            onLayout={[Function]}
            responderCallback={
              Object {
                "onMoveShouldSetPanResponder": [Function],
                "onMoveShouldSetPanResponderCapture": [Function],
                "onPanResponderGrant": [Function],
                "onPanResponderMove": [Function],
                "onPanResponderRelease": [Function],
                "onPanResponderTerminationRequest": [Function],
                "onShouldBlockNativeResponder": [Function],
                "onStartShouldSetPanResponder": [Function],
                "onStartShouldSetPanResponderCapture": [Function],
              }
            }
            style={
              Object {
                "height": 200,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "width": "100%",
              }
            }
            testID="PINCH_ZOOM_CONTAINER"
          >
            <Image
              onLoad={[Function]}
              resizeMode="contain"
              source={
                Object {
                  "uri": "https://images.unsplash.com/photo-1587628736664-fdc50efb57b7?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
                }
              }
              style={
                Object {
                  "backgroundColor": "#fff",
                  "height": 200,
                  "width": "100%",
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Array [
            Object {
              "maxWidth": 730,
              "minHeight": 200,
              "paddingBottom": 10,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 10,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
              },
            ]
          }
        >
          Image 4
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontSize": 14,
              },
            ]
          }
        >
          This line is for the description of the image
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#CCC",
                "height": 200,
                "width": "100%",
              },
              Object {
                "overflow": "hidden",
              },
            ]
          }
        >
          <View
            onLayout={[Function]}
            responderCallback={
              Object {
                "onMoveShouldSetPanResponder": [Function],
                "onMoveShouldSetPanResponderCapture": [Function],
                "onPanResponderGrant": [Function],
                "onPanResponderMove": [Function],
                "onPanResponderRelease": [Function],
                "onPanResponderTerminationRequest": [Function],
                "onShouldBlockNativeResponder": [Function],
                "onStartShouldSetPanResponder": [Function],
                "onStartShouldSetPanResponderCapture": [Function],
              }
            }
            style={
              Object {
                "height": 200,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "width": "100%",
              }
            }
            testID="PINCH_ZOOM_CONTAINER"
          >
            <Image
              onLoad={[Function]}
              resizeMode="contain"
              source={
                Object {
                  "uri": "https://p.bigstockphoto.com/eIdTXLbqQilMs9xbjvcs_bigstock-Aerial-View-Of-Sandy-Beach-Wit-256330393.jpg",
                }
              }
              style={
                Object {
                  "backgroundColor": "#fff",
                  "height": 200,
                  "width": "100%",
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Array [
            Object {
              "maxWidth": 730,
              "minHeight": 200,
              "paddingBottom": 10,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 10,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
              },
            ]
          }
        >
          Image 5
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontSize": 14,
              },
            ]
          }
        >
          This line is for the description of the image
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#CCC",
                "height": 200,
                "width": "100%",
              },
              Object {
                "overflow": "hidden",
              },
            ]
          }
        >
          <View
            onLayout={[Function]}
            responderCallback={
              Object {
                "onMoveShouldSetPanResponder": [Function],
                "onMoveShouldSetPanResponderCapture": [Function],
                "onPanResponderGrant": [Function],
                "onPanResponderMove": [Function],
                "onPanResponderRelease": [Function],
                "onPanResponderTerminationRequest": [Function],
                "onShouldBlockNativeResponder": [Function],
                "onStartShouldSetPanResponder": [Function],
                "onStartShouldSetPanResponderCapture": [Function],
              }
            }
            style={
              Object {
                "height": 200,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "width": "100%",
              }
            }
            testID="PINCH_ZOOM_CONTAINER"
          >
            <Image
              onLoad={[Function]}
              resizeMode="contain"
              source={
                Object {
                  "uri": "https://avatars2.githubusercontent.com/u/7970947?v=3&s=460",
                }
              }
              style={
                Object {
                  "backgroundColor": "#fff",
                  "height": 200,
                  "width": "100%",
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </ScrollView>
</View>
`;
